File: apps/web/lib/types/project-materials.ts
// ── Existing material CRUD types ──────────────────────────────────────────────

export type TrackingType = 'stock' | 'work'

export const TRACKING_TYPE_LABELS: Record<TrackingType, string> = {
  stock: 'Materijal',
  work:  'Radna aktivnost',
}

export interface MaterialListItem {
  id: string
  material_name: string
  material_code: string | null
  planned_quantity: number
  used_quantity: number
  available_quantity: number
  unit: string
  source: string | null
  active: boolean
  tracking_type: TrackingType
  created_at: string
  updated_at: string
}

export interface CreateMaterialPayload {
  material_name: string
  material_code?: string | null
  planned_quantity: number
  unit: string
  tracking_type: TrackingType
}

export interface UpdateMaterialPayload {
  material_name: string
  material_code?: string | null
  planned_quantity: number
  unit: string
  tracking_type: TrackingType
}

// ── Phase 6.5 wizard import types ────────────────────────────────────────────

/** Internal field names the backend understands. */
export type ImportField =
  | 'material_name'
  | 'quantity'
  | 'unit'
  | 'unit_price'
  | 'total_price'
  | 'category'
  | 'note'
  | 'position'
  | 'ignore'

/** Human-readable labels for each internal field. */
export const FIELD_LABELS: Record<ImportField, string> = {
  material_name: 'Naziv materijala',
  quantity: 'Količina',
  unit: 'Jedinica mjere',
  unit_price: 'Jedinična cijena',
  total_price: 'Ukupna cijena',
  category: 'Kategorija',
  note: 'Napomena',
  position: 'Pozicija / Rbr.',
  ignore: 'Ignoriraj',
}

/** Confidence thresholds for badge display — matches backend thresholds. */
export function confidenceLabel(conf: number): 'high' | 'medium' | 'low' | 'none' {
  if (conf >= 0.85) return 'high'
  if (conf >= 0.60) return 'medium'
  if (conf > 0) return 'low'
  return 'none'
}

// Step 1 — analyze response

export interface AnalyzeHeaderSuggestion {
  index: number
  original: string
  normalized: string
  suggested_field: ImportField | null
  confidence: number
  confidence_label: 'high' | 'medium' | 'low' | 'none'
  reasons: string[]
}

export interface AnalyzeSampleRow {
  row_number: number
  values: string[]
}

export interface ImportAnalyzeResponse {
  import_job_id: string
  sheet_name: string
  header_row_number: number
  headers: AnalyzeHeaderSuggestion[]
  sample_rows: AnalyzeSampleRow[]
}

// Step 2 — wizard preview

/** User-defined mapping: {originalExcelHeader: ImportField} */
export type ImportMapping = Record<string, ImportField>

export interface WizardPreviewRow {
  row_number: number
  material_name: string
  quantity: number
  unit: string
  unit_price: number | null
  total_price: number | null
  category: string | null
  note: string | null
  position: string | null
  source_row_numbers: number[]
  row_type: string
  warnings: string[]
  errors: string[]
  status: 'valid' | 'invalid'
}

export interface WizardPreviewSummary {
  total_rows: number
  valid_rows: number
  invalid_rows: number
  total_scanned: number
  skipped_categories: number
  skipped_continuations: number
  skipped_subtotals: number
}

export interface WizardPreviewResponse {
  rows: WizardPreviewRow[]
  summary: WizardPreviewSummary
}

// Step 3 — confirm

export interface ConfirmRow {
  row_number: number
  position?: string | null
  material_name: string
  quantity: number
  unit: string
  unit_price?: number | null
  total_price?: number | null
  category?: string | null
  note?: string | null
  include: boolean
}

export interface WizardConfirmRowError {
  row_number: number
  errors: string[]
}

export interface WizardConfirmResponse {
  imported_count: number
  skipped_count: number
  failed_count: number
  errors: WizardConfirmRowError[]
}

// ── Role helpers ──────────────────────────────────────────────────────────────

export function canImportMaterials(role: string): boolean {
  return role === 'direktor' || role === 'inzenjer'
}

export function canManageMaterials(role: string): boolean {
  return role === 'direktor' || role === 'inzenjer'
}

File: services/api/dto/project_materials.go
package dto

import "time"

type MaterialListItem struct {
	ID                string    `json:"id"`
	MaterialName      string    `json:"material_name"`
	MaterialCode      *string   `json:"material_code"`
	PlannedQuantity   float64   `json:"planned_quantity"`
	UsedQuantity      float64   `json:"used_quantity"`
	AvailableQuantity float64   `json:"available_quantity"`
	Unit              string    `json:"unit"`
	Source            *string   `json:"source"`
	Active            bool      `json:"active"`
	TrackingType      string    `json:"tracking_type"`
	CreatedAt         time.Time `json:"created_at"`
	UpdatedAt         time.Time `json:"updated_at"`
}

type CreateMaterialRequest struct {
	MaterialName    string  `json:"material_name"    binding:"required"`
	MaterialCode    *string `json:"material_code"`
	PlannedQuantity float64 `json:"planned_quantity" binding:"min=0"`
	Unit            string  `json:"unit"             binding:"required"`
	TrackingType    string  `json:"tracking_type"` // "stock" (default) or "work"
}

type UpdateMaterialRequest struct {
	MaterialName    string  `json:"material_name"    binding:"required"`
	MaterialCode    *string `json:"material_code"`
	PlannedQuantity float64 `json:"planned_quantity" binding:"min=0"`
	Unit            string  `json:"unit"             binding:"required"`
	TrackingType    string  `json:"tracking_type"` // "stock" (default) or "work"
}

// ── Legacy import types (kept for internal use) ───────────────────────────────

type ImportPreviewRow struct {
	RowNumber    int     `json:"row_number"`
	MaterialName string  `json:"material_name"`
	MaterialCode *string `json:"material_code"`
	Quantity     float64 `json:"quantity"`
	Unit         string  `json:"unit"`
	Status       string  `json:"status"` // valid | invalid
	ErrorMessage *string `json:"error_message"`
}

// ── Phase 6.5 wizard import types ────────────────────────────────────────────

// Step 1 — analyze: detect headers and suggest mappings

type AnalyzeHeaderSuggestion struct {
	Index           int      `json:"index"`
	Original        string   `json:"original"`
	Normalized      string   `json:"normalized"`
	SuggestedField  *string  `json:"suggested_field"`  // nil if confidence < 0.60
	Confidence      float64  `json:"confidence"`       // 0.0–1.0
	ConfidenceLabel string   `json:"confidence_label"` // "high" | "medium" | "low" | "none"
	Reasons         []string `json:"reasons"`          // human-readable explanation
}

type AnalyzeSampleRow struct {
	RowNumber int      `json:"row_number"`
	Values    []string `json:"values"` // cell values in column order
}

type ImportAnalyzeResponse struct {
	ImportJobID     string                    `json:"import_job_id"`
	SheetName       string                    `json:"sheet_name"`
	HeaderRowNumber int                       `json:"header_row_number"` // 1-based
	Headers         []AnalyzeHeaderSuggestion `json:"headers"`
	SampleRows      []AnalyzeSampleRow        `json:"sample_rows"`
}

// Step 2 — wizard preview: parse using user-selected mapping

type WizardPreviewRow struct {
	RowNumber        int      `json:"row_number"`
	MaterialName     string   `json:"material_name"`
	Quantity         float64  `json:"quantity"`
	Unit             string   `json:"unit"`
	UnitPrice        *float64 `json:"unit_price"`
	TotalPrice       *float64 `json:"total_price"`
	Category         *string  `json:"category"`
	Note             *string  `json:"note"`
	Position         *string  `json:"position"`
	SourceRowNumbers []int    `json:"source_row_numbers"`
	RowType          string   `json:"row_type"`
	Warnings         []string `json:"warnings"`
	Errors           []string `json:"errors"`
	Status           string   `json:"status"` // valid | invalid
}

type WizardPreviewSummary struct {
	TotalRows            int `json:"total_rows"`
	ValidRows            int `json:"valid_rows"`
	InvalidRows          int `json:"invalid_rows"`
	TotalScanned         int `json:"total_scanned"`
	SkippedCategories    int `json:"skipped_categories"`
	SkippedContinuations int `json:"skipped_continuations"`
	SkippedSubtotals     int `json:"skipped_subtotals"`
}

type WizardPreviewResponse struct {
	Rows    []WizardPreviewRow   `json:"rows"`
	Summary WizardPreviewSummary `json:"summary"`
}

// Step 3 — confirm: import (optionally edited) rows

// WizardConfirmRow represents one row sent from the frontend during confirm.
// The director may have edited field values before confirming.
type WizardConfirmRow struct {
	RowNumber    int      `json:"row_number"`
	Position     *string  `json:"position"`
	MaterialName string   `json:"material_name"`
	Quantity     float64  `json:"quantity"`
	Unit         string   `json:"unit"`
	UnitPrice    *float64 `json:"unit_price"`
	TotalPrice   *float64 `json:"total_price"`
	Category     *string  `json:"category"`
	Note         *string  `json:"note"`
	Include      bool     `json:"include"`
}

type WizardConfirmRowError struct {
	RowNumber int      `json:"row_number"`
	Errors    []string `json:"errors"`
}

type WizardConfirmResponse struct {
	ImportedCount int                     `json:"imported_count"`
	SkippedCount  int                     `json:"skipped_count"`
	FailedCount   int                     `json:"failed_count"`
	Errors        []WizardConfirmRowError `json:"errors"`
}
